server/src/test/java/org/elasticsearch/tasks/CancellableTasksTrackerTests.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.hamcrest.Matchers.not;
+
+public class CancellableTasksTrackerTests extends ESTestCase {
+
+    private static class TestTask {
+        private final Thread actionThread;
+        private final Thread watchThread;
+        private final Thread concurrentRemoveThread;
+
+        // 0 == before put, 1 == during put, 2 == after put, before remove, 3 == during remove, 4 == after remove
+        private final AtomicInteger state = new AtomicInteger();
+        private final boolean concurrentRemove = randomBoolean();
+
+        TestTask(Task task, String item, CancellableTasksTracker<String> tracker, Runnable awaitStart) {
+            if (concurrentRemove) {
+                concurrentRemoveThread = new Thread(() -> {
+                    awaitStart.run();
+
+                    for (int i = 0; i < 10; i++) {
+                        if (3 <= state.get()) {
+                            final String removed = tracker.remove(task);
+                            if (removed != null) {
+                                assertSame(item, removed);
+                            }
+                        }
+                    }
+                });
+            } else {
+                concurrentRemoveThread = new Thread(awaitStart);
+            }
+
+            actionThread = new Thread(() -> {
+                awaitStart.run();
+
+                state.incrementAndGet();
+                tracker.put(task, item);
+                state.incrementAndGet();
+
+                state.incrementAndGet();
```

Comment:
I am not entirely sure we would ever see state==2 in other threads, I would think it would be ok for the JVM to optimize it away.

This is possibly even true for the entire `actionThread` since there is no synchronization with the watch threads here. The watch threads may only see the the final output. I would advocate a yield should help, but that is outside JMM I think.

Reply:
Ah, an earlier iteration had some assertions in between these calls. In theory you're right that it would be legitimate for the compiler to collapse these tests to something fairly trivial, but in practice they do cover many interleavings and the other threads do indeed observe `state==2` cases. Just try adding bugs :) I'll add a `yield` anyway.